extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java

```diff
@@ -0,0 +1,266 @@
+     * doesn't exists
+     *
+     * @param filePath the path to the file you want
+     * @return a {@link FtpFileAttributes}
+     * @throws IllegalArgumentException if the {@code filePath} doesn't exists
+     */
+    protected FtpFileAttributes getExistingFile(String filePath)
+    {
+        return getFile(filePath, true);
+    }
+
+    /**
+     * Obtains a {@link FtpFileAttributes} for the given {@code filePath}
+     * by using the {@link FTPClient#mlistFile(String)} FTP command
+     *
+     * @param filePath the path to the file you want
+     * @return a {@link FtpFileAttributes} or {@code null} if it doesn't exists
+     */
+    public FtpFileAttributes getFile(String filePath)
+    {
+        return getFile(filePath, false);
+    }
+
+    protected abstract FtpFileAttributes getFile(String filePath, boolean requireExistence);
+
+    /**
+     * {@inheritDoc}
+     */
+    protected boolean exists(Path path)
+    {
+        return getFile(path.toString()) != null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected Path getBasePath()
+    {
+        return Paths.get(getCurrentWorkingDirectory());
+    }
+
+    /**
+     * Changes the current working directory to the given {@code path}
+     *
+     * @param path the {@link Path} to which you wish to move
+     * @throws IllegalArgumentException if the CWD could not be changed
+     */
+    protected void changeWorkingDirectory(Path path)
+    {
+        changeWorkingDirectory(path.toString());
+    }
+
+    /**
+     * Changes the current working directory to the given {@code path}
+     *
+     * @param path the path to which you wish to move
+     * @throws IllegalArgumentException if the CWD could not be changed
+     */
+    protected void changeWorkingDirectory(String path)
+    {
+        if (!tryChangeWorkingDirectory(path))
+        {
+            throw new IllegalArgumentException(format("Could not change working directory to '%s'. Path doesn't exists or is not a directory", path.toString()));
+        }
+        LOGGER.debug("working directory changed to {}", path);
+    }
+
+    /**
+     * Attempts to change the current working directory.
+     * If it was not possible (for example, because it doesn't exists),
+     * it returns {@code false}
+     *
+     * @param path the path to which you wish to move
+     * @return {@code true} if the CWD was changed. {@code false} otherwise
+     */
+    protected abstract boolean tryChangeWorkingDirectory(String path);
+
+    /**
+     * Template method that renames the file at {@code filePath} to {@code newName}.
+     * <p>
+     * This method performs path resolution and validation and eventually delegates
+     * into {@link #doRename(String, String)}, in which the actual renaming implementation
+     * is.
+     *
+     * @param filePath the path of the file to be renamed
+     * @param newName  the new name
+     */
+    protected void rename(String filePath, String newName)
```

Comment:
Should this method be final to make sure all the extensions have the same behaviour?